### ISSUE TITLE
rsa_pss_rsae_sha256 signatures from rsassa-pss SPKI certificates are accepted

### ISSUE THREAD

**Issue by tomato42** [openssl/openssl#6611]
When OpenSSL is configured to disable TLSv1.3 support and enable just rsa_pss_rsae_sha256 signature algorithm, it will accept that signature made with a certificate with rsassa-pss SPKI.

### Reproducer:
compile gnutls from commit cc1e19fbacc (the reproducer depends on gnutls bug, so it needs the buggy version; other versions may work but this is verified to misbehave):
   <details><summary>compilation instructions</summary><p>

1. install gnutls dependencies (see its [README.md](https://gitlab.com/gnutls/gnutls/blob/master/README.md))
1. download the code:
   ```
   git clone git@gitlab.com:gnutls/gnutls.git
   git checkout cc1e19fbacc
   ```
1. configure
   ```
   cd gnutls
   make autoreconf
   ./configure --disable-doc
   ```
1. compile
   ```
   make -j4
   ```
   </p></details>

start gnutls server:
```
cd gnutls/doc/credentials
./gnutls-http-serv \
--priority NORMAL:-VERS-ALL:+VERS-TLS1.3:+VERS-TLS1.2:+DHE-PSK:+PSK \
-p 4433 -a -d 6
```

connect with openssl (tested 358ffa05cd3a0888):
```
openssl s_client -connect localhost:4433 -no_tls1_3 \
-sigalgs rsa_pss_rsae_sha256 \
-CAfile gnutls/doc/credentials/x509/ca.pem
```

   <details><summary>openssl output</summary><p>

```
CONNECTED(00000005)
depth=1 CN = GnuTLS Test CA
verify return:1
depth=0 
verify return:1
---
Certificate chain
 0 s:
   i:CN = GnuTLS Test CA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDWTCCAhGgAwIBAgIIWYl7hABP8u0wDQYJKoZIhvcNAQELBQAwGTEXMBUGA1UE
AxMOR251VExTIFRlc3QgQ0EwIBcNMTcwODA4MDg1MTE2WhgPOTk5OTEyMzEyMzU5
NTlaMAAwggEgMAsGCSqGSIb3DQEBCgOCAQ8AMIIBCgKCAQEAndjwZ1p/m9BbCDGx
vn/yvqINoivwaCwrCQ94/ckTUhw+sE4n2MDU23tSxf/Ac8XMFUQg9M6W6RUGsG48
W5gUs616kJVBRRtdQ7zQWsOfK4BxqN02Aq7RSAXy284sgdcP71nl9MR77/DDCdYk
UnWPu2N+mtnFfrPOT3TuAU6WZS38vCzz+qevnYFaAvvbU7th9cAEWDlaIPo7fQNx
8dC9ccVVk3nRaIitrFaLs3y0Y36eXDsLkR0g9qm6RjgHjVVWjhPPAb96SBj4LjMN
KgHRA6NtIUWB9tyyMrwcAaAp8hTZwFYjLS2tkJV0pYlfWvQSjl2I1swHrKNkheKX
R3L3eQIDAQABo4GNMIGKMAwGA1UdEwEB/wQCMAAwFAYDVR0RBA0wC4IJbG9jYWxo
b3N0MBMGA1UdJQQMMAoGCCsGAQUFBwMBMA8GA1UdDwEB/wQFAwMHgAAwHQYDVR0O
BBYEFGp+ppXHKlMtWeuML0b98cYOIdspMB8GA1UdIwQYMBaAFE1Wt2oAWPFnkvSm
dVUbjlMBA+/PMA0GCSqGSIb3DQEBCwUAA4IBMQBiV5IoE/3e/YjP7FVyjXGzp8Kz
Zalj5fByCAMIwQ2oiWYsPYegNvJzUs42QMZ/AdCZdpXjTeFxPoN4T9pO+65ONGBL
X2mPCRu8oB2q9BPoOTV2ENVfLJHQa507ouZsC3/7NePZW+PlPM774Yxdhbhbj+4E
J2Onl7yQeHrXLgnbO8GzDHfjvG3Z6Zmrv80YcfGSXPtSr2vKMlXTc7yU+OD1s685
JjiCtIw2o5UmewJO4QpPPD2wcqwIQCXcCR2bVFZhf0/6nhfNQSFEK4qhJbOr2EY8
zesxVU2YWGCDKluoFxwEEHTpmpNUDmW8lYN3Dh8Qd2luKRQaGZUKc9Ve+M9iAzu6
2FmF6uGPfI0eu13eIe/Mas4zvFgYB43U5wWUGgNrvRASN9cyqVBqDIcLeuLj
-----END CERTIFICATE-----
subject=

issuer=CN = GnuTLS Test CA

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 1698 bytes and written 247 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 630BA71B5232768E8325FFE59489AF38C15FA576B9E48B9281D517352FA617CA
    Session-ID-ctx: 
    Master-Key: 89ED1054B7ED725BE873CC47DAAA1387ADF85B4FB960B1E75A5DE135F7A8D3686670DA1936C7A49FCBD5471C4FAAA687
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 21600 (seconds)
    TLS session ticket:
    0000 - f8 6b 96 cd 41 15 3a 47-99 bc 61 4c 9c a5 11 c2   .k..A.:G..aL....
    0010 - 3c 2c 56 a5 6e ca a1 01-8c 9d 2d a2 c7 f8 9d 0f   <,V.n.....-.....
    0020 - 01 20 d4 c9 bb 24 bf 57-da 60 3c 01 8c b7 47 2e   . ...$.W.`<...G.
    0030 - d8 a8 e4 63 2e 34 7d 56-3f b3 92 49 95 0e fd 37   ...c.4}V?..I...7
    0040 - 3f d9 cd 49 ea 6a 8b 48-ca db 0d 38 24 98 2a 0e   ?..I.j.H...8$.*.
    0050 - 32 bc e3 7d 1b f8 db d0-c5 40 33 cd 9f 2f 5d d7   2..}.....@3../].
    0060 - ba bb 04 9b b0 71 7e 6b-92 be 55 57 82 68 c6 10   .....q~k..UW.h..
    0070 - 23 ba cb 97 21 d5 93 43-37 b7 6f 3c eb 3b f7 87   #...!..C7.o<.;..
    0080 - ea 1b a6 da c4 e8 3b 99-4c 63 c9 ac b9 52 55 b4   ......;.Lc...RU.
    0090 - 86 4b 42 cc b9 2c bf 67-6e 2b 1a 6e 9f c3 29 be   .KB..,.gn+.n..).
    00a0 - fe 2b da 3e b8 29 f7 f3-63 b7 bc 26 b2 0a 6d 72   .+.>.)..c..&..mr
    00b0 - 0b 0a 60 4f 84 f5 5a 88-22 78 00 38 2a a0 8f a1   ..`O..Z."x.8*...
    00c0 - 09 f4 6a 32 6a 95 97 d9-f7 80 ad c3 c1 31 4a fe   ..j2j........1J.
    00d0 - 39 39 20 aa 75 a6 6c 26-71 12 e1 7d 16 50 4f d3   99 .u.l&q..}.PO.
    00e0 - 89 fb b8 21 4e e0 38 20-29 a1 37 77 08 6a 0b b5   ...!N.8 ).7w.j..
    00f0 - b8 df f3 aa b7 56 84 06-e2 34 5e 4c 2c ce 26 67   .....V...4^L,.&g
    0100 - 77 23 cc 5d 5e 04 93 bc-fc c4 7b 4e 89 ce 59 bd   w#.]^.....{N..Y.
    0110 - 0f 9d f8 b9 69 d4 a7 13-d3 bc d4 85 dc e9 a3 d5   ....i...........
    0120 - d6 be 0e bc a1 71 41 f1-5b fd 35 76 3b 2d 67 19   .....qA.[.5v;-g.
    0130 - 0a 87 9d 5a 3e 08 b5 48-e6 7e fa 57 1a e5 94 a6   ...Z>..H.~.W....
    0140 - a5 5f bb 4b e6 b4 89 a5-dd e5 ef 83 75 09 a3 4f   ._.K........u..O
    0150 - 09 e3 f7 e3 16 c8                                 ......

    Start Time: 1530197657
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: yes
---
DONE
```
   </p></details>

Note that the certificate presented has rsassa-pss SPKI:

   <details><summary>decoded certificate</summary><p>

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 6451823748056085229 (0x59897b84004ff2ed)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = GnuTLS Test CA
        Validity
            Not Before: Aug  8 08:51:16 2017 GMT
            Not After : Dec 31 23:59:59 9999 GMT
        Subject: 
        Subject Public Key Info:
            Public Key Algorithm: rsassaPss
                RSA-PSS Public-Key: (2048 bit)
                Modulus:
                    00:9d:d8:f0:67:5a:7f:9b:d0:5b:08:31:b1:be:7f:
                    f2:be:a2:0d:a2:2b:f0:68:2c:2b:09:0f:78:fd:c9:
                    13:52:1c:3e:b0:4e:27:d8:c0:d4:db:7b:52:c5:ff:
                    c0:73:c5:cc:15:44:20:f4:ce:96:e9:15:06:b0:6e:
                    3c:5b:98:14:b3:ad:7a:90:95:41:45:1b:5d:43:bc:
                    d0:5a:c3:9f:2b:80:71:a8:dd:36:02:ae:d1:48:05:
                    f2:db:ce:2c:81:d7:0f:ef:59:e5:f4:c4:7b:ef:f0:
                    c3:09:d6:24:52:75:8f:bb:63:7e:9a:d9:c5:7e:b3:
                    ce:4f:74:ee:01:4e:96:65:2d:fc:bc:2c:f3:fa:a7:
                    af:9d:81:5a:02:fb:db:53:bb:61:f5:c0:04:58:39:
                    5a:20:fa:3b:7d:03:71:f1:d0:bd:71:c5:55:93:79:
                    d1:68:88:ad:ac:56:8b:b3:7c:b4:63:7e:9e:5c:3b:
                    0b:91:1d:20:f6:a9:ba:46:38:07:8d:55:56:8e:13:
                    cf:01:bf:7a:48:18:f8:2e:33:0d:2a:01:d1:03:a3:
                    6d:21:45:81:f6:dc:b2:32:bc:1c:01:a0:29:f2:14:
                    d9:c0:56:23:2d:2d:ad:90:95:74:a5:89:5f:5a:f4:
                    12:8e:5d:88:d6:cc:07:ac:a3:64:85:e2:97:47:72:
                    f7:79
                Exponent: 65537 (0x10001)
                No PSS parameter restrictions
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name: 
                DNS:localhost
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
            X509v3 Key Usage: critical
                Digital Signature
            X509v3 Subject Key Identifier: 
                6A:7E:A6:95:C7:2A:53:2D:59:EB:8C:2F:46:FD:F1:C6:0E:21:DB:29
            X509v3 Authority Key Identifier: 
                keyid:4D:56:B7:6A:00:58:F1:67:92:F4:A6:75:55:1B:8E:53:01:03:EF:CF

    Signature Algorithm: sha256WithRSAEncryption
         62:57:92:28:13:fd:de:fd:88:cf:ec:55:72:8d:71:b3:a7:c2:
         b3:65:a9:63:e5:f0:72:08:03:08:c1:0d:a8:89:66:2c:3d:87:
         a0:36:f2:73:52:ce:36:40:c6:7f:01:d0:99:76:95:e3:4d:e1:
         71:3e:83:78:4f:da:4e:fb:ae:4e:34:60:4b:5f:69:8f:09:1b:
         bc:a0:1d:aa:f4:13:e8:39:35:76:10:d5:5f:2c:91:d0:6b:9d:
         3b:a2:e6:6c:0b:7f:fb:35:e3:d9:5b:e3:e5:3c:ce:fb:e1:8c:
         5d:85:b8:5b:8f:ee:04:27:63:a7:97:bc:90:78:7a:d7:2e:09:
         db:3b:c1:b3:0c:77:e3:bc:6d:d9:e9:99:ab:bf:cd:18:71:f1:
         92:5c:fb:52:af:6b:ca:32:55:d3:73:bc:94:f8:e0:f5:b3:af:
         39:26:38:82:b4:8c:36:a3:95:26:7b:02:4e:e1:0a:4f:3c:3d:
         b0:72:ac:08:40:25:dc:09:1d:9b:54:56:61:7f:4f:fa:9e:17:
         cd:41:21:44:2b:8a:a1:25:b3:ab:d8:46:3c:cd:eb:31:55:4d:
         98:58:60:83:2a:5b:a8:17:1c:04:10:74:e9:9a:93:54:0e:65:
         bc:95:83:77:0e:1f:10:77:69:6e:29:14:1a:19:95:0a:73:d5:
         5e:f8:cf:62:03:3b:ba:d8:59:85:ea:e1:8f:7c:8d:1e:bb:5d:
         de:21:ef:cc:6a:ce:33:bc:58:18:07:8d:d4:e7:05:94:1a:03:
         6b:bd:10:12:37:d7:32:a9:50:6a:0c:87:0b:7a:e2:e3
```
   </p></details>

**Comment by tomato42** [comment]
filed in gnutls under https://gitlab.com/gnutls/gnutls/issues/500

**Comment by mattcaswell** [comment]
Hmmm...I get a compilation failure trying to build that gnutls version:

````
verify.c:40:10: fatal error: supported_exts.h: No such file or directory
 #include "supported_exts.h"
          ^~~~~~~~~~~~~~~~~~
compilation terminated.
````

**Comment by mattcaswell** [comment]
Seems to be a problem with dependency ordering. Dropping the "-j4" solves it.

**Comment by tomato42** [comment]
@mattcaswell are you sure you have all the dependencies installed?

**Comment by nmav** [comment]
It requires gperf to be installed. README.md has the list of packages to be installed.

**Comment by mattcaswell** [comment]
Fix in #6732.
